### PR TITLE
Isolate Function Logic from Hosting Project

### PR DESCRIPTION
**Pull request by wsugarman**
## Description
- Like the API and Web project, split the existing function project into two: operations and function
- The new `Microsoft.Health.Dicom.Operations` contains the logic for our Azure functions, like `Microsoft.Health.Dicom.Functions` contains the actual hosting and defines the more immutable function metadata for the HTTP triggers
- Rename `Microsoft.Health.Dicom.Functions.UnitTests` to `Microsoft.Health.Dicom.Operations.UnitTests`
- Update namespaces
- Refactor service registration for reuse

## Related issues
[AB#86438](https://microsofthealth.visualstudio.com/Health/_workitems/edit/86438)

## Testing
Ad-hoc and via pipeline

### CODE REVIEW

File: src/Microsoft.Health.Dicom.Functions/Management/DurableClientProxyDecorator.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+using Microsoft.Azure.WebJobs.Extensions.DurableTask;
+using Microsoft.Azure.WebJobs.Extensions.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Health.Dicom.Operations.Management;
+
+namespace Microsoft.Health.Dicom.Functions.Management
+{
+    public class DurableClientProxyDecorator
+    {
+        private readonly DurableClientProxy _proxy;
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Functions/Management/DurableClientProxyDecorator.cs:22]: nit: can probably seal this

File: src/Microsoft.Health.Dicom.Functions/Indexing/ReindexDurableFunctionDecorator.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+namespace Microsoft.Health.Dicom.Functions.Indexing
+{
+    public sealed class ReindexDurableFunctionDecorator
+    {
+        private readonly ReindexDurableFunction _function;
+
+        public ReindexDurableFunctionDecorator(ReindexDurableFunction function)
+            => _function = EnsureArg.IsNotNull(function, nameof(function));
+
+        [FunctionName(nameof(StartReindexingInstancesAsync))]
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Functions/Indexing/ReindexDurableFunctionDecorator.cs:26]: As you can see, I am creating the HTTP functions in this project by using the various attributes to specify auth, path, etc

File: src/Microsoft.Health.Dicom.Functions/Microsoft.Health.Dicom.Functions.csproj [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
   <PropertyGroup>
     <AzureFunctionsVersion>v3</AzureFunctionsVersion>
+    <FunctionsInDependencies>true</FunctionsInDependencies>
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Functions/Microsoft.Health.Dicom.Functions.csproj:5]: Many of the functions are now present in the dependency package `Microsoft.Health.Dicom.Operations`

File: src/Microsoft.Health.Dicom.Operations/Registration/ServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -11,49 +11,63 @@
...
-                .Bind(
-                    configuration
-                        .GetSection(DicomFunctionsConfiguration.SectionName)
-                        .GetSection(sectionName),
-                    x => x.BindNonPublicProperties = bindNonPublicProperties)
-                .ValidateDataAnnotations();
+            services.RegisterModule<ServiceModule>(new FeatureConfiguration { EnableExtendedQueryTags = true });
 
-            return services;
+            return new DicomFunctionsBuilder(services
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Operations/Registration/ServiceCollectionExtensions.cs:39]: This file has seen a number of refactors to allow for reuse between here and a PAAS equivalent.

File: Microsoft.Health.Dicom.sln [hunk 5]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio Version 16
-VisualStudioVersion = 16.0.28729.10
+# Visual Studio Version 17
+VisualStudioVersion = 17.0.31825.309
```
Comment (wsugarman) [inline Microsoft.Health.Dicom.sln:4]: This version won't be out of preview until next week!

**Review by pengchen0692 — APPROVED** [review]

**Pull request merged**